agent/acs/handler/acs_handler.go

```diff
@@ -52,37 +53,47 @@ var SequenceNumber = utilatomic.NewIncreasingInt64(1)
 // StartSession creates a session with ACS and handles requests using the passed
 // in arguments.
 func StartSession(containerInstanceArn string, credentialProvider credentials.AWSCredentialProvider, cfg *config.Config, taskEngine engine.TaskEngine, ecsclient api.ECSClient, stateManager statemanager.StateManager, acceptInvalidCert bool) error {
-	backoff := utils.NewSimpleBackoff(time.Second, 1*time.Minute, 0.2, 2)
-	return utils.RetryWithBackoff(backoff, func() error {
-		acsEndpoint, err := ecsclient.DiscoverPollEndpoint(containerInstanceArn)
-		if err != nil {
-			log.Error("Unable to discover poll endpoint", "err", err)
-			return err
-		}
-
-		url := AcsWsUrl(acsEndpoint, cfg.Cluster, containerInstanceArn, taskEngine)
-
-		client := acsclient.New(url, cfg.AWSRegion, credentialProvider, acceptInvalidCert)
-		defer client.Close()
-
-		client.AddRequestHandler(payloadMessageHandler(client, cfg.Cluster, containerInstanceArn, taskEngine, ecsclient, stateManager))
-		client.AddRequestHandler(heartbeatHandler(client))
-
-		updater.AddAgentUpdateHandlers(client, cfg, stateManager, taskEngine)
-
-		err = client.Connect()
-		if err != nil {
-			log.Error("Error connecting to ACS: " + err.Error())
-			return err
+	backoff := utils.NewSimpleBackoff(time.Second, 2*time.Minute, 0.2, 2)
+	for {
+		acsError := func() error {
```

Comment:
I have a minor aversion to anonymous functions unless there is a compelling reason (like the desire to close over things outside the function) as I think they're a bit harder to read.  Feel free to leave it as-is if you think it's appropriate.